### PR TITLE
ObstacleAvoidance: fix condition for target wp skipping

### PR DESCRIPTION
**Pull request by Luka-Filipovic**
## Describe problem solved by this pull request
Condition for skipping current target waypoint is wrong. 
```
// if the vehicle projected position on the line previous-target is past the target waypoint,
// increase the target acceptance radius such that navigator will update the triplets
```
The idea here is to skip the target waypoint if it's already overpassed. (e.g. due to obstacle avoidance)
The distance from the closest point on track to the previous waypoint is compared to the distance from the target waypoint to the previous waypoint. If the former is greater than the latter, current target waypoint is skipped. The problem occurs if the distance between previous and target waypoint is shorter than previous waypoint's acceptance radius. 

Example: previous waypoint acceptance radius is 2 meters, all waypoints are colinear:
`vehicle_pos <-- 1.9 meters --> prev_wp <-- 1 meter --> target_wp <-- 10 meters --> next_wp`

It's obvious that the target waypoint will be skipped, but so will the next_wp and every that comes after, as long as they are in the same direction and the condition is met.

This is dangerous, especially if the mission is planned to avoid obstacles.

## Describe your solution
Condition should require vectors `prev_wp -> target_wp` and `prev_wp -> closest` to have the same direction. It is easily checked if the dot product is greater than 0.

### CODE REVIEW

**Comment by Luka-Filipovic** [comment]
> It seems like this part handles the case you are mentioning. But I am not 100% sure if it will actually handle the case properly.

That internal triplet is only relevant for the FlightTask and trajectory generation. Acceptance radius modified here in ObstacleAvoidance is later used in navigator/mission_block to check if the current waypoint is reached.
 
> The only valid test case would be a SITL I believe. Could you perform a SITL test with the corner case you mentioned? @Luka-Filipovic

This bug was identified during the real tests, whenever there were colinear waypoints separated by no more than acc rad, waypoint were skipped in advance. I'll try to find some time to test it on vanilla px4

**Comment by Luka-Filipovic** [comment]
> Could you perform a SITL test with the corner case you mentioned?

Here's one extreme scenario where almost the whole mission gets skipped:
Screenshot from 2023-08-07 16-24-13

After the change:
Screenshot from 2023-08-07 16-26-18

Some info on the testing setup:
We initially discovered this bug on a v1.10 version of PX4, and the logic that's causing the bug hasn't changed since. However, there is a "newer" change in https://github.com/PX4/PX4-Autopilot/commit/4c35b3aadc62407bb5b5403714fb3dbfeb5e9826 that limits the acceptance radius update to non rotary-wings vehicles. I'm not sure why that is the case, we find this waypoint skip feature useful for our drone system. 
So to prove the point, I also had to remove that limit since I'm testing it in quadcopter SITL.
```
--- a/src/modules/navigator/navigator_main.cpp
+++ b/src/modules/navigator/navigator_main.cpp
@@ -1215,9 +1215,7 @@ float Navigator::get_acceptance_radius()
        const position_controller_status_s &pos_ctrl_status = _position_controller_status_sub.get();
 
        // for fixed-wing and rover, return the max of NAV_ACC_RAD and the controller acceptance radius (e.g. navigation switch distance)
-       if (_vstatus.vehicle_type != vehicle_status_s::VEHICLE_TYPE_ROTARY_WING
-           && PX4_ISFINITE(pos_ctrl_status.acceptance_radius) && pos_ctrl_status.timestamp != 0) {
-
+       if (pos_ctrl_status.timestamp != 0) {
                acceptance_radius = math::max(acceptance_radius, pos_ctrl_status.acceptance_radius);
        }
``` 
But the same issue should still be there for other vehicle types because it is just a small math bug.

**Comment by dusan19** [comment]
@mrivi we think this is still relevant. Could we get your opinion and review?
